src/coreclr/jit/lsra.cpp

```diff
@@ -5922,9 +5922,62 @@ void LinearScan::allocateRegisters()
             assert(lclVarInterval->isLocalVar);
             if (refType == RefTypeUpperVectorSave)
             {
-                if ((lclVarInterval->physReg == REG_NA) ||
+                assert(currentInterval->recentRefPosition == &currentRefPosition);
+
+                // For a given RefTypeUpperVectorSave, there should be a matching RefTypeUpperVectorRestore
+                // If not, probably, this was an extra one we added conservatively and should not need a register.
```

Comment:
I don't understand this comment.

If we're saving something across a callsite, we must restore it. Otherwise, the save was never needed in the first place and it can simply be bypassed.

Reply:
> Otherwise, the save was never needed in the first place and it can simply be bypassed.

That's the whole point. Basically, earlier we were saving only for live variables, but we found out that sometimes the liveness information can be inaccurate and so we are saving for all the `bbLiveIn` and `bbDefs`, but later, we try to find out and see which one of the saves are unnecessary and bypass it by the virtue of this and `IsExtraUpperVectorSave`

Reply:
Can we not just fix liveness to ensure that other presumptions aren't incorrect?

Seemingly if liveness is wrong, any other phase of the JIT could also be doing things incorrectly.